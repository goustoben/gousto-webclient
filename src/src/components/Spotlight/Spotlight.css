.spotlight {
  position: absolute;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;
  background-color: #000000;
  opacity: 0.4;
  cursor: none;
  z-index: 100;
}
