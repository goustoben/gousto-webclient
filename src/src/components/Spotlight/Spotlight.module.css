@value TrueBlack from 'styles/colors.module.css';

.spotlight {
  position: absolute;
  top: 0;
  left: 0;
  min-height: 100vh;
  width: 100vw;
  background-color: TrueBlack;
  opacity: 0.4;
  z-index: 400;
}
