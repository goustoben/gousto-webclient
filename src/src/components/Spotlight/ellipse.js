export const getEllipse = ({ originX, originY, radius }) => {
  const n = 360,
    baseAngle = 2 * Math.PI / n

  let angle, x, y, steps = ''

  for(var i = 0; i <= n; i++) {
    angle = i * baseAngle
    x = Math.cos(angle)
    y = Math.sin(angle)

    steps += `${(originX + radius * x)}px ${(originY - radius * y)}px, `
  }

  return steps
}
