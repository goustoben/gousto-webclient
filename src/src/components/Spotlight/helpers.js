export const getDocumentHeight = () => {
  const { body, documentElement } = document

  return Math.max(
    body.scrollHeight,
    body.offsetHeight,
    documentElement.clientHeight,
    documentElement.scrollHeight,
    documentElement.offsetHeight,
  )
}

export const getEllipse = ({ originX, originY, radius, accuracy = 1 }) => {
  const n = 360 * accuracy,
    baseAngle = 2 * Math.PI / n

  let angle, x, y, steps = ''

  for (var i = 0; i <= n; i++) {
    angle = i * baseAngle
    x = Math.cos(angle).toFixed(3)
    y = Math.sin(angle).toFixed(3)

    steps += `${(originX + radius * x)}px ${(originY - radius * y)}px, `
  }

  return steps
}
