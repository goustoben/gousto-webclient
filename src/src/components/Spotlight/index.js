export { Spotlight } from './Spotlight'
export { SpotlightWrapper } from './SpotlightWrapper'
