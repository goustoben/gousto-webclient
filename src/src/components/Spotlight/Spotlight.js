import React from 'react'
import PropTypes from 'prop-types'

import { getEllipse } from './ellipse'
import css from './Spotlight.css'

const RADIUS = 100
const accuracy = 0.3

const Spotlight = ({ x, y }) => {
  const steps = getEllipse({
    originX: x,
    originY: y,
    radius: RADIUS,
    accuracy,
  })

  return (
    <div
      className={css.spotlight}
      style={{
        clipPath: `polygon(
          0% 0%,
          100% 0%,
          100% ${y}px,
          ${steps}
          100% ${y}px,
          100% 100%,
          0% 100%
        )`,
      }}
    >
    </div>
  )
}

Spotlight.defaultProps = {
  x: 0,
  y: 0,
}

Spotlight.propTypes = {
  x: PropTypes.number,
  y: PropTypes.number,
}

export {
  Spotlight,
}
