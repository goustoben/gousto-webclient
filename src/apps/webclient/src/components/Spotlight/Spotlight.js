import React from 'react'
import PropTypes from 'prop-types'
import { onEnter } from 'utils/accessibility'
import { getDocumentHeight, getEllipse } from './helpers'
import css from './Spotlight.css'

const RADIUS = 100
const accuracy = 0.3

const Spotlight = ({ x, y, onClick }) => {
  const steps = getEllipse({
    originX: x,
    originY: y,
    radius: RADIUS,
    accuracy,
  })

  return (
    <div
      role="button"
      tabIndex={0}
      data-testing="spotlight-overlay"
      className={css.spotlight}
      style={{
        height: getDocumentHeight(),
        clipPath: `polygon(
          0% 0%,
          100% 0%,
          100% ${y}px,
          ${steps}
          100% ${y}px,
          100% 100%,
          0% 100%
        )`,
        '-webkit-clip-path': `polygon(
          0% 0%,
          100% 0%,
          100% ${y}px,
          ${steps}
          100% ${y}px,
          100% 100%,
          0% 100%
        )`,
      }}
      onClick={onClick}
      onKeyDown={onEnter(onClick)}
    />
  )
}

Spotlight.propTypes = {
  x: PropTypes.number,
  y: PropTypes.number,
  onClick: PropTypes.func,
}

Spotlight.defaultProps = {
  x: 0,
  y: 0,
  onClick: null,
}

export { Spotlight, RADIUS, accuracy }
