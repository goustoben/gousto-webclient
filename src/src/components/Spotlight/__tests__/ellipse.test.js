import { getEllipse } from 'Spotlight/ellipse'

describe('getEllipse', () => {
  let accuracy
  const ellipseParams = {
    originX: 100,
    originY: 100,
    radius: 50,
  }

  describe('with an accuracy of 0.1', () => {
    beforeEach(() => {
      accuracy = 0.1
    })

    test('should generate clip-path co-ordinates', () => {
      expect(getEllipse({ ...ellipseParams, accuracy })).toMatchSnapshot()
    })
  })

  describe('with an accuracy of 0.2', () => {
    beforeEach(() => {
      accuracy = 0.2
    })

    test('should generate clip-path co-ordinates', () => {
      expect(getEllipse({ ...ellipseParams, accuracy })).toMatchSnapshot()
    })
  })

  describe('with an accuracy of 0.3', () => {
    beforeEach(() => {
      accuracy = 0.3
    })

    test('should generate clip-path co-ordinates', () => {
      expect(getEllipse({ ...ellipseParams, accuracy })).toMatchSnapshot()
    })
  })
})
