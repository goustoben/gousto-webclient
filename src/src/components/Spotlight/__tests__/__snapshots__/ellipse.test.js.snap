// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEllipse with an accuracy of 0.1 should generate clip-path co-ordinates 1`] = `"150px 100px, 149.25px 91.3px, 147px 82.9px, 143.3px 75px, 138.3px 67.85px, 132.15px 61.7px, 125px 56.7px, 117.1px 53px, 108.7px 50.75px, 100px 50px, 91.3px 50.75px, 82.9px 53px, 75px 56.7px, 67.85px 61.7px, 61.7px 67.85px, 56.7px 75px, 53px 82.9px, 50.75px 91.3px, 50px 100px, 50.75px 108.7px, 53px 117.1px, 56.7px 125px, 61.7px 132.15px, 67.85px 138.3px, 75px 143.3px, 82.9px 147px, 91.3px 149.25px, 100px 150px, 108.7px 149.25px, 117.1px 147px, 125px 143.3px, 132.15px 138.3px, 138.3px 132.15px, 143.3px 125px, 147px 117.1px, 149.25px 108.7px, 150px 100px, "`;

exports[`getEllipse with an accuracy of 0.2 should generate clip-path co-ordinates 1`] = `"150px 100px, 149.8px 95.65px, 149.25px 91.3px, 148.3px 87.05px, 147px 82.9px, 145.3px 78.85px, 143.3px 75px, 140.95px 71.3px, 138.3px 67.85px, 135.35px 64.65px, 132.15px 61.7px, 128.7px 59.050000000000004px, 125px 56.7px, 121.15px 54.699999999999996px, 117.1px 53px, 112.95px 51.7px, 108.7px 50.75px, 104.35px 50.2px, 100px 50px, 95.65px 50.2px, 91.3px 50.75px, 87.05px 51.7px, 82.9px 53px, 78.85px 54.699999999999996px, 75px 56.7px, 71.3px 59.050000000000004px, 67.85px 61.7px, 64.65px 64.65px, 61.7px 67.85px, 59.050000000000004px 71.3px, 56.7px 75px, 54.699999999999996px 78.85px, 53px 82.9px, 51.7px 87.05px, 50.75px 91.3px, 50.2px 95.65px, 50px 100px, 50.2px 104.35px, 50.75px 108.7px, 51.7px 112.95px, 53px 117.1px, 54.699999999999996px 121.15px, 56.7px 125px, 59.050000000000004px 128.7px, 61.7px 132.15px, 64.65px 135.35px, 67.85px 138.3px, 71.3px 140.95px, 75px 143.3px, 78.85px 145.3px, 82.9px 147px, 87.05px 148.3px, 91.3px 149.25px, 95.65px 149.8px, 100px 150px, 104.35px 149.8px, 108.7px 149.25px, 112.95px 148.3px, 117.1px 147px, 121.15px 145.3px, 125px 143.3px, 128.7px 140.95px, 132.15px 138.3px, 135.35px 135.35px, 138.3px 132.15px, 140.95px 128.7px, 143.3px 125px, 145.3px 121.15px, 147px 117.1px, 148.3px 112.95px, 149.25px 108.7px, 149.8px 104.35px, 150px 100px, "`;

exports[`getEllipse with an accuracy of 0.3 should generate clip-path co-ordinates 1`] = `"150px 100px, 149.9px 97.1px, 149.65px 94.2px, 149.25px 91.3px, 148.65px 88.45px, 147.9px 85.65px, 147px 82.9px, 145.9px 80.2px, 144.7px 77.55px, 143.3px 75px, 141.75px 72.5px, 140.1px 70.15px, 138.3px 67.85px, 136.35px 65.69999999999999px, 134.3px 63.65px, 132.15px 61.7px, 129.85px 59.9px, 127.5px 58.25px, 125px 56.7px, 122.45px 55.3px, 119.8px 54.1px, 117.1px 53px, 114.35px 52.1px, 111.55px 51.35px, 108.7px 50.75px, 105.8px 50.35px, 102.9px 50.1px, 100px 50px, 97.1px 50.1px, 94.2px 50.35px, 91.3px 50.75px, 88.45px 51.35px, 85.65px 52.1px, 82.9px 53px, 80.2px 54.1px, 77.55px 55.3px, 75px 56.7px, 72.5px 58.25px, 70.15px 59.9px, 67.85px 61.7px, 65.69999999999999px 63.65px, 63.65px 65.69999999999999px, 61.7px 67.85px, 59.9px 70.15px, 58.25px 72.5px, 56.7px 75px, 55.3px 77.55px, 54.1px 80.2px, 53px 82.9px, 52.1px 85.65px, 51.35px 88.45px, 50.75px 91.3px, 50.35px 94.2px, 50.1px 97.1px, 50px 100px, 50.1px 102.9px, 50.35px 105.8px, 50.75px 108.7px, 51.35px 111.55px, 52.1px 114.35px, 53px 117.1px, 54.1px 119.8px, 55.3px 122.45px, 56.7px 125px, 58.25px 127.5px, 59.9px 129.85px, 61.7px 132.15px, 63.65px 134.3px, 65.69999999999999px 136.35px, 67.85px 138.3px, 70.15px 140.1px, 72.5px 141.75px, 75px 143.3px, 77.55px 144.7px, 80.2px 145.9px, 82.9px 147px, 85.65px 147.9px, 88.45px 148.65px, 91.3px 149.25px, 94.2px 149.65px, 97.1px 149.9px, 100px 150px, 102.9px 149.9px, 105.8px 149.65px, 108.7px 149.25px, 111.55px 148.65px, 114.35px 147.9px, 117.1px 147px, 119.8px 145.9px, 122.45px 144.7px, 125px 143.3px, 127.5px 141.75px, 129.85px 140.1px, 132.15px 138.3px, 134.3px 136.35px, 136.35px 134.3px, 138.3px 132.15px, 140.1px 129.85px, 141.75px 127.5px, 143.3px 125px, 144.7px 122.45px, 145.9px 119.8px, 147px 117.1px, 147.9px 114.35px, 148.65px 111.55px, 149.25px 108.7px, 149.65px 105.8px, 149.9px 102.9px, 150px 100px, "`;
